plugins/package-managers/bazel/src/main/kotlin/Bazel.kt

```diff
@@ -0,0 +1,309 @@
+                project = Project(
+                    id = Identifier(
+                        type = managerName,
+                        name = moduleMetadata.module?.name ?: "root",
+                        version = moduleMetadata.module?.version.orEmpty(),
+                        namespace = ""
+                    ),
+                    definitionFilePath = VersionControlSystem.getPathInfo(definitionFile).path,
+                    authors = emptySet(),
+                    declaredLicenses = emptySet(),
+                    vcs = projectVcs,
+                    vcsProcessed = projectVcs,
+                    homepageUrl = "",
+                    scopeDependencies = scopes
+                ),
+                packages = packages.toSet()
+            )
+        )
+    }
+
+    private fun getPackages(
+        graph: Set<PackageReference>,
+        bmrClient: BazelModuleRegistryClient,
+        seen: MutableSet<Identifier> = mutableSetOf()
+    ): Set<Package?> {
+        val result = mutableSetOf<Package?>()
+
+        graph.forEach { ref ->
+            if (ref.id !in seen) {
+                seen.add(ref.id)
+
+                if (ref.id.name != "<root>") {
+                    result.add(packageFromPackageReference(ref, bmrClient))
+                }
+                result.addAll(getPackages(ref.dependencies, bmrClient, seen))
+            }
+        }
+
+        return result
+    }
+
+    private fun packageFromPackageReference(ref: PackageReference, bmrClient: BazelModuleRegistryClient): Package? {
+        val metadata: ModuleMetadata
+        try {
+            metadata = runBlocking { bmrClient.getModuleMetadata(ref.id.name) }
+        } catch (e: HttpException) {
+            logger.warn { "Failed to fetch metadata for Bazel module ${ref.id.name}: $e" }
+            return null
+        }
+
+        val sourceInfo: ModuleSourceInfo
+        try {
+            sourceInfo = runBlocking { bmrClient.getModuleSourceInfo(ref.id.name, ref.id.version) }
+        } catch (e: HttpException) {
+            logger.warn { "Failed to fetch source info for Bazel module ${ref.id.name}@${ref.id.version}: $e" }
+            return null
+        }
+
+        return Package(
+            id = ref.id,
+            declaredLicenses = emptySet(),
+            description = "",
+            homepageUrl = metadata.homepage?.toString().orEmpty(),
+            binaryArtifact = RemoteArtifact.EMPTY,
+            sourceArtifact = sourceInfo.remoteArtifact(),
+            vcs = toVcsInfo(metadata)
+        )
+    }
+
+    private fun getDependencyGraph(
+        projectDir: File,
+        depDirectives: Map<String, BazelDepDirective>,
+        lockFile: LockFile?
+    ): Set<Scope> {
+        if (lockFile != null) {
+            return parseLockFileGraph(lockFile.moduleDepGraph, depDirectives, managerName)
+        } else {
+            val modGraphProcess = run("mod", "graph", "--output", "json", workingDir = projectDir)
```

Comment:
Do edges in this Graph represent version constraints or actual (resolved) dependencies? (Asking as I'm a bit reminded of `GoMod` looking at the `cli` command)

Reply:
I don't know. AFAIK bzlmod uses the same dependency resolution approach as Go mod, i.e. finding the lowest version that satisfies all constraints. But I don't know in what scenario it actually picks versions. [The documentation on `bazel_dep`](https://bazel.build/rules/lib/globals/module#bazel_dep) just says:

> The version of the module to be added as a direct dependency.

It also says the default of this attribute is `''`, which to me suggests that the version is optional. However, when I try that I get an error:

```
$ cat MODULE.bazel
bazel_dep(name = "glog", repo_name = "com_github_google_glog")
bazel_dep(name = "googletest", version = "1.14.0", repo_name = "com_google_googletest", dev_dependency = True)

$ bazel mod graph --output json
ERROR: in module dependency chain <root> -> glog@_: bad bazel_dep on module 'glog' with no version. Did you forget to specify a version, or a non-registry override?. Type 'bazel help mod' for syntax and help.
```